File: src/java/es/logongas/fpempresa/modelo/comun/Usuario.java
/**
 * FPempresa Copyright (C) 2014 Lorenzo González
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package es.logongas.fpempresa.modelo.comun;

import es.logongas.ix3.security.model.User;
import es.logongas.ix3.core.annotations.Label;
import es.logongas.ix3.security.authentication.Principal;
import javax.validation.constraints.AssertTrue;
import javax.validation.constraints.NotNull;
import org.hibernate.event.spi.PostLoadEvent;
import org.hibernate.event.spi.PostLoadEventListener;
import org.hibernate.validator.constraints.Email;
import org.hibernate.validator.constraints.NotBlank;

/**
 *
 * @author Lorenzo
 */
public class Usuario extends User  implements PostLoadEventListener,Principal {
    
    @Email
    @NotBlank
    @Label("Correo electrónico")
    private String eMail;
    
    @NotBlank
    private String nombre;
    
    @Label("1º Apellido")
    @NotBlank
    private String ape1;
    
    @Label("2º Apellido")
    private String ape2;
    
    private byte[] foto;
    
    @Label("Contraseña")
    private String password;

    @NotNull
    @Label("Tipo de usuario")
    private TipoUsuario tipoUsuario;
    
    
    public Usuario() {
        this.tipoUsuario=TipoUsuario.TITULADO;
    }

    @Override
    public String toString() {
        return name;
    }

    /**
     * @return the eMail
     */
    public String geteMail() {
        return eMail;
    }

    /**
     * @param eMail the eMail to set
     */
    public void seteMail(String eMail) {
        this.eMail = eMail;
        this.login=this.eMail;
    }
    /**
     * @return the eMail
     */
    public String getEMail() {
        return eMail;
    }

    /**
     * @param eMail the eMail to set
     */
    public void setEMail(String eMail) {
        this.eMail = eMail;
        this.login=this.eMail;
    }
       
    /**
     * @return the nombre
     */
    public String getNombre() {
        return nombre;
    }

    /**
     * @param nombre the nombre to set
     */
    public void setNombre(String nombre) {
        this.nombre = nombre;
        this.name=getCalculateName();
    }

    /**
     * @return the ape1
     */
    public String getApe1() {
        return ape1;
    }

    /**
     * @param ape1 the ape1 to set
     */
    public void setApe1(String ape1) {
        this.ape1 = ape1;
        this.name=getCalculateName();        
    }

    /**
     * @return the ape2
     */
    public String getApe2() {
        return ape2;
    }

    /**
     * @param ape2 the ape2 to set
     */
    public void setApe2(String ape2) {
        this.ape2 = ape2;
        this.name=getCalculateName();        
    }

    /**
     * @return the foto
     */
    public byte[] getFoto() {
        return foto;
    }

    /**
     * @param foto the foto to set
     */
    public void setFoto(byte[] foto) {
        this.foto = foto;
    }

    /**
     * @return the password
     */
    public String getPassword() {
        return password;
    }

    /**
     * @param password the password to set
     */
    public void setPassword(String password) {
        this.password = password;
    }

    /**
     * @return the tipoUsuario
     */
    public TipoUsuario getTipoUsuario() {
        return tipoUsuario;
    }

    /**
     * @param tipoUsuario the tipoUsuario to set
     */
    public void setTipoUsuario(TipoUsuario tipoUsuario) {
        this.tipoUsuario = tipoUsuario;
    }
    
    
    private String getCalculateName() {
        return nombre + " " + ape1 + (ape2 != null ? " " + ape2 : "");
    }
    
    @Override
    public void onPostLoad(PostLoadEvent ple) {
        Usuario usuario=(Usuario)ple.getEntity();
        //Nunca se retorna el Hash de la contraseña
        usuario.setPassword(null);

    }   
    
    
    /*
    @AssertTrue(message = "El registro está deshabilitado")
    @Label("")
    private boolean isProhibidoNuevoUsuario() {
        return false;
    }
    
    @AssertTrue(message = "Solo se permite registrar titulados")
    @Label("")
    private boolean isSoloPermitidoTitulados() {
        if (this.tipoUsuario==TipoUsuario.TITULADO) {
            return true;
        } else {
            return false;
        }
    } 
    */
    
}

File: web/common/domain/comun/usuario/Usuario.js
/**
 *   FPempresa
 *   Copyright (C) 2020  Lorenzo González
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Affero General Public License as
 *   published by the Free Software Foundation, either version 3 of the
 *   License, or (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Affero General Public License for more details.
 *
 *   You should have received a copy of the GNU Affero General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
"use strict";

angular.module("common").config(['richDomainProvider', function (richDomain) {

        richDomain.addEntityTransformer("Usuario", ['schemaEntities', function (schemaEntities) {
                var Usuario = {
                    getNombreCompleto: function () {
                        return this.nombre + (this.apellidos ? " " + this.apellidos : "");
                    },
                    getEstadoUsuarioDescription: function () {
                        return schemaEntities.getSchemaProperty(this.$propertyPath + ".estadoUsuario").getValueDescription(this.estadoUsuario);
                    },
                    getTipoUsuarioDescription: function () {
                        return schemaEntities.getSchemaProperty(this.$propertyPath + ".tipoUsuario").getValueDescription(this.tipoUsuario);
                    },
                    $validators: [
                        {
                            label: 'Confirmar Contraseña',
                            message: 'No es igual a la contraseña',
                            executeInActions: ['INSERT'],
                            stopOnFail:true,
                            rule: function () {
                                if (this.password === this.confirmPassword) {
                                    return true;
                                } else {
                                    return false;
                                }
                            }
                        },
                        {
                            label: "Terminos de uso y privacidad",
                            message: 'Debe aceptar los terminos de uso y la política de privacidad',
                            executeInActions: ['INSERT'],
                            stopOnFail:true,
                            rule: function () {
                                if (this.aceptarCondicionesPolitica === true) {
                                    return true;
                                } else {
                                    return false;
                                }
                            }
                        },
                        {
                            label: "Exactitud, veracidad y ofertas",
                            message: 'Debe aceptar que los centros educativos puedan comprobar los datos académicos así como conocer en que ofertas me he inscrito',
                            executeInActions: ['INSERT'],
                            stopOnFail:true,
                            rule: function () {
                                if (this.tipoUsuario === 'TITULADO') {
                                    if (this.aceptarVerificarTitulo === true) {
                                        return true;
                                    } else {
                                        return false;
                                    }
                                } else {
                                    return true;
                                }
                            }
                        },
                        {
                            label: "Envio de correos",
                            message: 'Si no marcas esta opción no te llegarán por correo las ofertas de empleo',
                            executeInActions: ['INSERT'],
                            rule: function () {
                                if (this.tipoUsuario!=="TITULADO") {
                                    return true;
                                }
                                
                                if (this.__avisadoDebeAceptarEnvioCorreos===true) {
                                    return true;
                                }
                                
                                this.__avisadoDebeAceptarEnvioCorreos=true;
                                
                                if (this.aceptarEnvioCorreos === true) {
                                    return true;
                                } else {
                                    return false;
                                }
                            }
                        }
                    ]
                };

                return function (object, propertyPath) {
                    object.$propertyPath = propertyPath;
                    angular.extend(object, Usuario);
                };
            }]);

    }]);